ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixParticipantMetrics.java

```diff
@@ -16,55 +16,47 @@
+import java.util.Collections;
+import java.util.Map;
 import java.util.concurrent.atomic.AtomicInteger;
 
 
 /**
  * Metrics for {@link HelixParticipant} to monitor partition state transitions.
  */
 class HelixParticipantMetrics {
-  final AtomicInteger bootstrapCount = new AtomicInteger();
-  final AtomicInteger standbyCount = new AtomicInteger();
-  final AtomicInteger leaderCount = new AtomicInteger();
-  final AtomicInteger inactiveCount = new AtomicInteger();
   final AtomicInteger offlineCount = new AtomicInteger();
-  final AtomicInteger errorStateCount = new AtomicInteger();
   // no need to record exact number of "dropped" partition, a counter to track partition-dropped events would suffice
   final Counter partitionDroppedCount;
 
-  HelixParticipantMetrics(MetricRegistry metricRegistry, String zkConnectStr) {
+  HelixParticipantMetrics(MetricRegistry metricRegistry, String zkConnectStr,
+      Map<String, ReplicaState> localPartitionAndState) {
     String zkSuffix = zkConnectStr == null ? "" : "-" + zkConnectStr;
-    Gauge<Integer> bootstrapPartitionCount = bootstrapCount::get;
+    Gauge<Integer> bootstrapPartitionCount =
+        () -> Collections.frequency(localPartitionAndState.values(), ReplicaState.BOOTSTRAP);
     metricRegistry.register(MetricRegistry.name(HelixParticipant.class, "bootstrapPartitionCount" + zkSuffix),
         bootstrapPartitionCount);
-    Gauge<Integer> standbyPartitionCount = standbyCount::get;
+    Gauge<Integer> standbyPartitionCount =
+        () -> Collections.frequency(localPartitionAndState.values(), ReplicaState.STANDBY);
     metricRegistry.register(MetricRegistry.name(HelixParticipant.class, "standbyPartitionCount" + zkSuffix),
         standbyPartitionCount);
-    Gauge<Integer> leaderPartitionCount = leaderCount::get;
+    Gauge<Integer> leaderPartitionCount =
+        () -> Collections.frequency(localPartitionAndState.values(), ReplicaState.LEADER);
     metricRegistry.register(MetricRegistry.name(HelixParticipant.class, "leaderPartitionCount" + zkSuffix),
         leaderPartitionCount);
-    Gauge<Integer> inactivePartitionCount = inactiveCount::get;
+    Gauge<Integer> inactivePartitionCount =
+        () -> Collections.frequency(localPartitionAndState.values(), ReplicaState.INACTIVE);
     metricRegistry.register(MetricRegistry.name(HelixParticipant.class, "inactivePartitionCount" + zkSuffix),
         inactivePartitionCount);
-    Gauge<Integer> offlinePartitionCount = offlineCount::get;
+    Gauge<Integer> offlinePartitionCount =
+        () -> Collections.frequency(localPartitionAndState.values(), ReplicaState.OFFLINE);
     metricRegistry.register(MetricRegistry.name(HelixParticipant.class, "offlinePartitionCount" + zkSuffix),
         offlinePartitionCount);
-    Gauge<Integer> errorStatePartitionCount = errorStateCount::get;
+    Gauge<Integer> errorStatePartitionCount =
+        () -> Collections.frequency(localPartitionAndState.values(), ReplicaState.ERROR);
```

Comment:
This definitely is cleaner logic wise.
My only concern is that the gauge readings will result in 6 full scans of the map every minute (or whatever frequency), which could get expensive if the number of replicas is large.  A possible optimization would be to do a single scan periodically and update local counters for each state, which the gauges could read as before.  But the performance benefit may be too small to justify this.

Reply:
Fair point, I will try to scan once and cache the result for other 5 metrics.